### ISSUE TITLE
FM4 oscillator

### ISSUE THREAD

**Issue by mkruselj** [surge-synthesizer/surge#3619]
As discussed in #1029 and suggested by @baconpaul, perhaps the request posed by that issue could be more easily solved with a new oscillator type, let's aptly call it FM4.

So which parameters would we require?

Algorithm (see [this image](https://user-images.githubusercontent.com/2393720/104370724-96f64080-551f-11eb-9fa3-13aa25fadbe4.png))
O1 Shape (all the stuff from Sine osc)
O1 Ratio (has Absolute and Extend range)
O1 Amount
O2 Shape
O2 Ratio
O2 Amount
O3 Shape
O3 Ratio
O3 Amount
O4 Shape
O4 Ratio
O4 Amount
Feedback

That's the bare minimum, 14 parameters. Would be awesome if we could also have Offset parameters for each operator, which adds a constant detune in Hz (4 more parameters). Minor thing, though, Although it _is_ found in TX81z...

**Comment by baconpaul** [comment]
One thing I was thinking last night: Since we have param space we could add a little ADSR envelope as well and have O2->O4 be modulatable by it so you don’t have to burn an LFO slot if you want to do internal FM swells and stuff? May be a bad idea but I worry we are going to run out of modulators (!!) if we aren’t careful with some of this new stuff

**Comment by mkruselj** [comment]
You mean as part of osc? Hrm, not a fan personally. I mean we have 12 LFOs, so to me the correct solution is #2334...

**Comment by magnetophon** [comment]
How would you feel about having an [FM formant oscillator](https://chrischafe.net/glitch-free-fm-vocal-synthesis/) in FM4?
It ads one more parameter per operator: formant frequency.
Of course we'd also need a way to select this type, maybe it can be one of the shapes?

With ratio = 0, you hear a mix of two sinewaves an octave apart.
The frequency of those oscillators is that of the note, but in the two octaves closest to the formant frequency.
As you tweak the formant, the octaves cross-fade into each other.
When you increase the ratio, the bandwidth increases as you hear the FM between the base frequency and the formants come in to play.

You can play with the parameters, hear the result and see it on a scope, [all in the browser](https://faustide.grame.fr/?autorun=1&voices=0&name=untitled&inline=Ly8tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLQovLyBGTSBmb3JtYW50IG9zY2lsYXRvcgovLy0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tCgpkZWNsYXJlIG5hbWUgCQkiRk1Wb3giOwpkZWNsYXJlIHZlcnNpb24JCSIxLjAiOwpkZWNsYXJlIGF1dGhvcgkJIkNocmlzIENoYWZlIjsKZGVjbGFyZSBsaWNlbnNlCQkiQlNEIjsKZGVjbGFyZSBjb3B5cmlnaHQJInN0ayI7CgppbXBvcnQoInN0ZGZhdXN0LmxpYiIpOwoKLy8gLS0tLS0tIFByb2dyYW0gQSBvZiAiR2xpdGNoIEZyZWUgRk0gVm9jYWwgU3ludGhlc2lzIiAobWlub3IgY29ycmVjdGlvbnMgZnJvbSBhcnRpY2xlKQoKLy8gLS0tLS0tIHRhYmxlIGxvb2t1cCBvc2NpbGxhdG9ycyAtLS0tLS8vCnRzIAk9IDEgPDwgMTY7CQkJLy8gdGFibGUgc2l6ZQpmcyAJPSBmbG9hdCh0cyk7CnRzMQk9IHRzKzE7CmN0IAk9ICgrKDEpfl8gKSAtIDE7CQkvLyBpbmNyZW1lbnRpbmcgY291bnRlciBmcm9tIDAKZmN0CT0gZmxvYXQoY3QpOwpzYyAJPSBmY3QqKDIqbWEuUEkpL2ZzOnNpbjsJCS8vIHNpbmUgdGFibGUgZm9yIGNhcnJpZXIKc20JPSBmY3QqKDIqbWEuUEkpL2ZzOnNpbjovKDIqbWEuUEkpOwkvLyBzaW5lIHRhYmxlIGZvciBtb2R1bGF0b3IKZGVjKHgpCT0geC1mbG9vcih4KTsJCQkvLyBmcmFjdGlvbmFsIHJlbWFpbmRlcgpwaGEoZikJPSBmL2Zsb2F0KG1hLlNSKTooKzpkZWMpIH4gXzsJLy8gZ2VuZXJhdGVzIGEgaW5kZXggc2lnbmFsCnRibCh0LHApPSBzMQogICAgICAgICAgLy8gdGJsKHQscCk9IHMxK2RlYyhmKSooczItczEpCQkvLyBsb29rcyB1cCBsaW5lYXJseSBpbnRlcnBvbGF0ZWQgdGFibGUgdmFsdWUKd2l0aCB7CiAgZiA9IHAqZnM7CiAgaSA9IGludChmKTptaW4odHMpOwogIHMxID0gcmR0YWJsZSh0czEsdCxpKTsKICBzMiA9IHJkdGFibGUodHMxLHQsaSsxKTsKfTsKCi8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8KLy8gICAgICBmb3JtYW50IGdlbmVyYXRvciB1c2luZyB1bmlmb3JtIChwaGFzZS1zeW5jaHJvbm91cykgb3NjaWxsYXRvcnMgICAgICAvLwovLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vLy8vCgovLyBzeW5jZWQgdG8gYW4gZXh0ZXJuYWwgcGhhc2VyICJpbmRleCIKLy8gdGhlIHBhcmFtZXRlciBlbyBzZXRzIHRoZSBwcm9wb3J0aW9uIG9mIGV2ZW4gYW5kIG9kZCBoYXJtb25pY3MuCmZ1cGhvU2xhdmVFdmVuT2RkKGluZGV4LGYwLGEsYixjLGVvKSA9IChldmVuK29kZCk6KihhKQkvLyBvdXRwdXRzIHRoZSBzdW0gb2YgYnJhY2tldGluZyBoYXJtb25pY3MKd2l0aCB7CQkJCQkvLyBmcm9tIGYwLCBhbXAsIGJhbmR3aWR0aCwgY2VudGVyIGZyZXEKICBva3QgPSB2c2xpZGVyKCJva3QiLCAwLCAtMiwgMiwgMSk7CiAgY2YgCT0gYy9mMDsKICAvLyBjZiAJPSBjLyhmMCoob2t0Om9jdGF2ZU11bHRpcGxpZXIpKTsKICBjaQk9IGludChmbG9vcihjZikpOwkJCS8vIGludGVnZXIgaGFybW9uaWMgYmVsb3cgY2VudGVyIGZyZXEKICBjaTEJPSBjaSsxOwkJCQkvLyBhbmQgYWJvdmUKICBpc0V2ZW49IGJhLmlmKCgoY2klMik8MSksMSwwKTsJLy8gYmVsb3cgaXMgZXZlbj8KICBlZiAJPSBiYS5pZihpc0V2ZW4sY2ksY2kxKTsJCS8vIHRoZW4gc2V0IGV2ZW4gaGFybW9uaWMgdG8gbG93ZXN0CiAgb2YgCT0gYmEuaWYoaXNFdmVuLGNpMSxjaSk7CQkvLyBhbmQgb2RkIGhhcm1vbmljIHRvIGhpZ2hlc3QKICBmcmFjCT0gY2YtY2k7CQkJLy8gZnJhY3Rpb25hbCBmcmVxdWVuY3kgcmVtYWluZGVyCiAgY29tcAk9IDEtZnJhYzsKICBvYSAJPSBiYS5pZihpc0V2ZW4sZnJhYyxjb21wKSplbzsJCS8vIG9kZCBoYXJtb25pYyBhbXBsaXR1ZGUKICBlYSAJPSBiYS5pZihpc0V2ZW4sY29tcCxmcmFjKSooKGVvKi0xKSsxKTsJCS8vIGV2ZW4gaGFybW9uaWMgYW1wbGl0dWRlCiAgLy8gcGggCT0gcGhhKGYwKTsJCQkvLyBpbmRleCBzaWduYWwgYXQgZnVuZGFtZW50YWwKICBtIAk9IHRibChzbSxpbmRleCk6KihiKTsJCS8vIG1vZHVsYXRvciBzaW5lIHNpZ25hbAogIGV2ZW4gCT0gZWE6Kih0Ymwoc2MsKGRlYyhlZippbmRleCttKSkpKTsgLy8gZXZlbiBoYXJtb25pYyBzaWduYWwgd2l0aCBwaGFzZSBtb2R1bGF0aW9uCiAgb2RkIAk9IG9hOioodGJsKHNjLChkZWMob2YqaW5kZXgrbSkpKSk7CS8vIG9kZCBoYXJtb25pYyBzaWduYWwKfTsKCmZ1bmQoZnJlcSxwaGFzZSxyZXNldCkgPSBvcy5sZl9zYXdwb3NfcGhhc2VfcmVzZXQoZnJlcSxwaGFzZSxyZXNldCk7CgoKcHJvY2VzcyA9CiBmdXBob1NsYXZlRXZlbk9kZChmdW5kKGZyZXEscGhhc2UscmVzZXQpLGZyZXEsMSxGTWluZGV4LGZvcm1hbnRGcmVxLGV2ZW5PZGQpCiA8OihfLF8pCjsKCgpmcmVxID0gaHNsaWRlcigiWzFdZnJlcSIsIDQ0MCwgMjAsIDI0MDAwLCAxKSA6c2kuc21vbzsKZm9ybWFudEZyZXEgPSBoc2xpZGVyKCJbMl1mb3JtYW50IGZyZXF1ZW5jeSIsIDQ0MCwgMjcuNSwgMzUyMCwgMC4xKTpzaS5zbW9vOwpGTWluZGV4ID0gaHNsaWRlcigiWzNdRk0gaW5kZXhbdG9vbHRpcDogRk0gaW5kZXggb2YgdGhlIGxvd2VzdCBvc2NpbGxhdG9yXSIsCTEsIDAsIDIwLCAwLjAwMSk6bWF4KDAuMDAwMDAwMDEpOnNpLnNtb287CmV2ZW5PZGQgPSBoc2xpZGVyKCJbNF1ldmVuT2RkIiwgMC41LCAwLCAxLCAwLjAwMSk6c2kuc21vbzsKcGhhc2UgPSBoc2xpZGVyKCJbNV0gcGhhc2UiLCAwLCAtNjQsIDY0LCAwLjAwMSkgOnNpLnNtb287CnJlc2V0ID0gYnV0dG9uKCJbNl1yZXNldCIpOwo%3D).

**Comment by magnetophon** [comment]
I would also love to have that offset parameter, it's cool for making harmonic sounds with a controlled amount of beating.

Also, how about a phase param per operator. One of my favorite FM tricks is having two sets of operators, one panned left and one right and the only difference between L and R is the relative phase of the modulators.

**Comment by mkruselj** [comment]
We're trying not to have a forest of parameters per oscillator. Keep it as simple as possible but still felxible, y'know.

**Comment by baconpaul** [comment]
Phase control is something folks always ask about though - worth pondering

**Comment by magnetophon** [comment]
Yeah, not having a forest of parameters per oscillator is probably wise.
My software always ends up having a gazilion features and no users that dear come close to it.  ;)

**Comment by mkruselj** [comment]
> Phase control is something folks always ask about though - worth pondering

Sure but maybe global phase for the whole osc rather than per operator...

**Comment by PopeKehoeIII** [comment]
there is definitely an argument to be made that FM4 should stick more or less to emulating the 4 op TX style synths while the sine osc or phase distortion osc (or 2 op + wavefolder "complex osc") becomes the "super operator" that is used with the main FM matrix between operators.

I also agree with mkruselj that putting modulators inside of oscillators seems messy compared to just adding more modulation options to the synth as a whole (making LFO/SLFO a toggle option for all LFOs instead of having fixed amounts of each, or just adding another row of modulation slots so we can just get a bunch more LFOs added)

Global phase for the whole osc is something that could be useful in most osc models though, especially if the goal is panning tricks since pan is not available per operator inside of FM2/3(/4?)

**Comment by mkruselj** [comment]
Except we also have a "pan per osc" issue we intend to tackle for 1.9. 😉

**Comment by PopeKehoeIII** [comment]
but I assume that's meant to be "per osc" and not "4 pan sliders in one osc so you have pan per operator inside of FM3"?

**Comment by magnetophon** [comment]
Fixed LCR panning like in the current mixer would be enough in most cases, no? you can fit 4 of those on place of one slider, I think.

Per operator panning opens up a lot of options for cool wide sounds.

**Comment by mkruselj** [comment]
Yeah I did mean pan per osc in the mixer, not pan per operator (that's way overkill IMO).

So there will be pan per osc in 1.9 (as a knob), but it will work in addiition to the filter routing selector (they're not the same thing). And we will have new and updated filter configs that are all stereo always so that unison can actually have stereo spread parameter and so on. When it rains, it pours!

**Comment by j5v** [comment]
Maybe we can have these features, and avoid the forest, by offering them only on the operators where they're most effective, and choosing neutral defaults in other operators.

When compared to the forest, it would make some decisions to limit creativity a bit, but capture the majority of useful sweet spots. UI grouping might have to be more obvious, as a user can't use the same visual map for all operators.

**Comment by mkruselj** [comment]
Since the idea is to also have the Algorithm parameter, this means that all 4 operators need to have identical parameter set.

**Comment by PopeKehoeIII** [comment]
Just commenting to say that if algorithm is a modulate-able param looking at the dx100 algos I don't see a reason there couldn't be interpolation between them (with snap to nearest toggle-able via context menu maybe?) by gradually lowering the amplitude of some connections and raising others - lower numbered operators never modulate higher numbered operators and operator 4 is the only one with feedback normally. May be too complicated but it could make algorithm a really fun modulation target.
